src/validators/model_fields.rs

```diff
@@ -372,8 +372,24 @@ impl Validator for ModelFieldsValidator {
             }
         }?;
 
+        let new_extra = match &self.extra_behavior {
+            ExtraBehavior::Allow => {
+                let non_extra_data = PyDict::new(py);
+                self.fields.iter().for_each(|f| {
+                    let popped_value = PyAny::get_item(new_data, &f.name).unwrap();
+                    new_data.del_item(&f.name).unwrap();
+                    non_extra_data.set_item(&f.name, popped_value).unwrap();
+                });
+                let new_extra = new_data.copy()?;
+                new_data.clear();
+                new_data.update(non_extra_data.as_mapping())?;
+                new_extra.to_object(py)
+            }
+            _ => py.None(),
+        };
+
         let fields_set: &PySet = PySet::new(py, &[field_name.to_string()])?;
-        Ok((new_data, py.None(), fields_set.to_object(py)).to_object(py))
+        Ok((new_data.to_object(py), new_extra, fields_set.to_object(py)).to_object(py))
```

Comment:
The return value of `validate_assignment` in `_pydantic_core.pyi` is `dict[str, Any]`, but here we have a 3-tuple. 

a) Should we change the return value of `validate_assignment` in `_pydantic_core.pyi`?
b) Should we change the definition of `validate_assignment` in `trait Validator` to return `(PyObject, PyObject, PyObject)`? This will enforce all the Rust implementation to behave correctly and avoid needing to go in and out of a Python tuple until we get to the top level.

EDIT based on what I see in the tests, I think the answer to both of these is "yes".

Reply:
> Should we change the definition of validate_assignment in trait Validator to return (PyObject, PyObject, PyObject)? This will enforce all the Rust implementation to behave correctly and avoid needing to go in and out of a Python tuple until we get to the top level.

Some historical perspective: `validate_assignment` and `validate` used to be the same thing, there was just a boolean flag being thrown around internally to differentiate the "mode". So the fact that the return is a `PyObject` instead of a tuple of `PyObject` is likely just an artifact of that original implementation.

Reply:
This validate_assignment is also used for dataclasses, and I think we need it to be a `dict[str, Any]` for that case. I updated the type hint to be a union. I agree this probably deserves some cleanup but I think it's not as straightforward as "always return a tuple".